Xplane & text output from tracy

protobird-git opened the pull request:
A draft PR to output xplane protobuf and text to stdout from tracy.

okkwon commented on build_tools/bazel/iree.bazelrc:
```diff
@@ -152,6 +152,7 @@ build:generic_clang --copt=-Wunused-comparison
 build:generic_clang --copt=-Wvla
 
 # Treat backrefs during linking as errors.
+build:generic_clang --linkopt=-fuse-ld=lld
```
I think this is supposed to be controlled since it is not supported in the Apple environment. See IREE_ENABLE_LLD.

protobird-git replied:
Oh, it's not meant to be included this PR. Will remove it.

protobird-git replied:
Removed

okkwon commented on tools/iree-tracy-print.cc:
```diff
@@ -0,0 +1,304 @@
+  }
+}
+
+void PrintWorker(const tracy::Worker& worker) {
```
I think we need a separate file for each output type, i.e., print and xplane. This would help us create a separate converting tool like `iree-tracy-convert`, where we can reuse the files.

protobird-git replied:
Done

okkwon commented on tools/iree-tracy-print.cc:
```diff
@@ -0,0 +1,304 @@
+  std::cout << "[TRACY]     CpuArch = " << ArchToString(worker.GetCpuArch())
+            << "\n";
+
+  std::cout << "[TRACY]   CPU Zones = " << worker.GetZoneCount() << "\n";
```
Maybe we can prefix with [TRACY_CPU] or [CPU] to make `grep` happy.

protobird-git replied:
Done

okkwon commented on tools/iree-tracy-print.cc:
```diff
@@ -0,0 +1,304 @@
+  std::cout << "[TRACY]   CPU Zones = " << worker.GetZoneCount() << "\n";
+  auto cpu_zones = GetTopZones(worker.GetSourceLocationZones(), 10);
+  for (const auto& z : cpu_zones) {
+    std::cout << "[TRACY]         " << GetZoneName(worker, z.source_location_id)
```
Same. Let's prefix with [TRACY_CPU] or just [CPU].

protobird-git replied:
Done

okkwon commented on tools/iree-tracy-print.cc:
```diff
@@ -0,0 +1,304 @@
+              << ", count=" << z.zone->zones.size() << "\n";
+  }
+
+  std::cout << "[TRACY]   GPU Zones = " << worker.GetGpuZoneCount() << "\n";
```
Same here and below. [TRACY_GPU] or [GPU].

protobird-git replied:
Done

okkwon commented on tools/iree-tracy-print.cc:
```diff
@@ -0,0 +1,304 @@
+
+  std::unique_ptr<tracy::Worker> worker;
+  // Build a worker from a tracy file,
+  if (argc_for_subprogram == 0) {
```
I think having a separate tool like `iree-prof-convert` is simpler. Also, then, we can use `iree-prof` for the current program name, which does:

1. help launch tracy profiler.
2. save the output as tracy profile data, text, and xplane.

In addition, we can provide `iree-prof-convert` to convert tracy to the other formats.

protobird-git replied:
Done

okkwon commented on tools/iree-tracy-print.cc:
I'd like to suggest `iree-prof` for the program name.

protobird-git replied:
Done

okkwon commented on tools/iree-tracy-print.cc:
```diff
@@ -0,0 +1,304 @@
+IREE_FLAG(string, output_xplane_file, "",
+          "Xplane file to write as the output of execution or conversion.");
+IREE_FLAG(int32_t, tracy_port, 18086, "TCP port number for tracy profiler.");
+
```
With the name change, we would need an output option for printing since pinting is a kind of output.

protobird-git replied:
Done

okkwon reviewed (changes requested):
What a progress! Thanks a lot. Left a few comments.

protobird-git commented:
Updated the PR. Here is the commit message of the last commit:

Split iree-tracy-print into multiple pieces.
    
    1) iree-prof-output[-xxx].h/.cc is to output tracy result in various
       ways according to command line flags.
    2) iree-prof is to run subcommand and output tracy result.
    3) iree-prof-convert is to read a tracy file and output result in
       various ways according to command line flags.
    4) Depends on abseil-cpp instead of iree/base to prepare
       google/iree-prof-tools.

Plan to update stdout format in a next PR.

okkwon commented on tools/CMakeLists.txt:
```diff
@@ -258,13 +258,59 @@ iree_cc_binary(
+    absl::strings
+    absl::time
+    tracy::worker
+    tsl::xplane
```
We may need to build the xplane-related conditionally, but that can be done later.

protobird-git replied:
Ack

okkwon commented on tools/CMakeLists.txt:
```diff
@@ -258,13 +258,59 @@ iree_cc_binary(
+    absl::log_initialize
+    absl::strings
+    tracy::worker
+    tsl::xplane
```
This dependency shouldn't be needed.

protobird-git replied:
Oops, done.

okkwon commented on tools/CMakeLists.txt:
```diff
@@ -258,13 +258,59 @@ iree_cc_binary(
+    absl::log
+    absl::log_initialize
     tracy::worker
     tsl::xplane
```
Same, this is handled by `iree-prof-output` so we don't need it here.

protobird-git replied:
Done

okkwon reviewed (approved):
Looks great! I left a few very minor comments.
